### ISSUE TITLE
Add a command to temporarily paste in different mode

### ISSUE THREAD

**Issue by anselmwang** [jglev/obsidian-paste-to-current-indentation#11]
Hi, jglev

Your plugin is so amazing! Before using your plugin, I find myself constantly following some patterns like "paste first, then select pasted content, then add block quote or code block, indent". This pattern is very tedious.

Thank you for creating the plugin!

After using for a while, I find a small inconvenience. When I want to temporarily paste in a different mode, currently I have to set mode, paste and revert mode. This is a bit annoying. Can we have a "Paste in Mode..." command that pop up the mode selection dialog and let us select a mode just for one time? maybe bind it to "Ctrl-Shift-V" or another shortcut.

I am happy to create a PR if you agree on this change.

Thanks,
Yu

**Comment by jglev** [comment]
Thank you for your kind words!

And I agree with your assessment about temporary commands being useful. When I wrote the current version, I had in mind that the plugin could be used with the ["QuickAdd" plugin's "Macro"](https://github.com/chhoumann/quickadd/blob/master/docs/Choices/MacroChoice.md) feature to bind switching and pasting in a specific mode to a hotkey. But that doesn't cover the case of switching back into / staying in the current mode.

I look forward to taking a look at your PR this weekend!

**Comment by anselmwang** [comment]
Thanks a lot for help! 

BTW, I am a little hesitant to learn "QuickAdd" although I have heard it a lot. With your introduction, I just install and give it a try. It looks amazing!

**Comment by anselmwang** [comment]
Hi, @jglev 

I have verified `paste-to-current-indentation` doesn't work with "QuickAdd" because "QuickAdd" will not trigger any paste event, so will not trigger code of `paste-to-current-indentation`.
Below is the implementation of QuickAdd's Paste command:
```js
  static async run(app: App) {
      const clipboard = await navigator.clipboard.readText();
      const activeView = EditorCommand.getActiveMarkdownView(app);

      if (!activeView) {
          log.logError("no active markdown view.");
          return;
      }

      activeView.editor.replaceSelection(clipboard);
  }
```

Thanks,
Yu

**Comment by jglev** [comment]
@anselmwang, ah, that's good to know about QuickAdd! I hadn't realized that it wasn't invoking a system paste command; thank you for your explanation!

I really like the approach that you've taken here! I do note a couple of things, trying the functionality so far:

- If I copy an image to the clipboard and then try to paste in Passthrough mode, the paste seems to be unsuccessful.
- "Paste as Code Block (Blockquote)" seems not to parse newlines -- hence, this text:

```
fdkjdsjlkdflkjds
   Test 2
      Test 4
```

...becomes this text when pasted:

```
> ```
> fdkjdsjlkdflkjds,   Test 2,      Test 4
> ```
```

Are those behaviors happening on your system, too?

**Comment by anselmwang** [comment]
# For the image
For the image, the correct way to paste is to set the paste mode to "Passthrough" and `Ctrl-V` to trigger paste. Only `Ctrl-v` can trigger paste event so triggering other plugins and Obsidian's internal paste hook.  

This is how I use `paste-to-current-indentation` these days. I always the paste mode to "Passthrough", paste link and images with `Ctrl-V` and call "Paste as ..." commands for other requirements.

I shall remove the "Paste as Passthrough" command because the command doesn't do what the name implies. And due to lack of "trigger paste event" of Obsidian, I guess it is not able to implement "Paste as Passthrough". I think the best solution is what you currently provide: the "Passsthrough" mode.

# For "Paste as Code Block (Blockquote)"
Firstly, I can't reproduce the problem.
- paste from raw text
![Animation1](https://user-images.githubusercontent.com/7410579/158217844-d686c62c-c528-479d-a32e-1025b548c1b6.gif)
- paste from html
![Animation](https://user-images.githubusercontent.com/7410579/158218288-11142835-469a-40f5-b3c4-0d3985624043.gif)

I have also verified copy from VSCode, the command also works.

The current implementation simply retrieves clipboard contents by `evt.clipboardData.getData` and I have verified this API is reliable. I guess the reason you see is because the source of the code put some rich text, the rich text somehow doesn't work with `clipboardContents = evt.clipboardData.getData("text");` API as in the below code
```javascript
        if (
          mode === Mode.Text ||
          mode === Mode.TextBlockquote ||
          mode === Mode.CodeBlock ||
          mode === Mode.CodeBlockBlockquote
        ) {
          clipboardContents = evt.clipboardData.getData("text");
        }
```

To debug this issue, you may download a clipboard tool. On Windows platform, I am using InsideClipboard. It tells me everything about clipboard content
![image](https://user-images.githubusercontent.com/7410579/158218992-77969df4-fc8a-4f24-a205-ade929dde4ea.png)

# A side note
The pasted code has added ",", this looks like a bug?

**Comment by jglev** [comment]
@anselmwang, thank you for your response! I'm writing just to confirm that I haven't sat down with the changes yet, but I am planning to do so!

> The pasted code has added ",", this looks like a bug?

I've noticed this recently, too, and agree that it's a bug.

**Comment by jglev** [comment]
@anselmwang, thank you again!

I am finding that in the latest changeset, the Markdown and Markdown-Blockquote modes fail to paste when plain text is on the clipboard. Thus, I'd like to look at this a bit more. I've reached out just now to others in Obsidian Discord to learn more about the `editor-paste` event and whether there might be an alternative feasible approach to addressing this Issue. So, I hope to have more of an update soon!

**Comment by anselmwang** [comment]
Thanks a lot for pushing this forward!

"Markdown and Markdown-Blockquote modes fail to paste when plain text is on the clipboard" is expected, When I set "Paste Mode: Markdown" and plain text is on the clipboard, "Ctrl-v" also fails.

This is because `evt.clipboardData.getData("text/html")` or any other method which tries to retrieve html will fail when there are only plain text in the clipboard. This is how clipboard designed. "html" content can be converted to "text" content but the other direction is not allowed.

**Comment by jglev** [comment]
You're welcome!

> This is because evt.clipboardData.getData("text/html") or any other method which tries to retrieve html will fail when there are only plain text in the clipboard. This is how clipboard designed.

This is true, but I meant that the change breaks existing functionality in the plugin. 

After learning on an Obsidian Discord channel that it is possible to trigger a paste event programmatically, I've published a [pre-release](https://github.com/jglev/obsidian-paste-to-current-indentation/releases/tag/2.2.0-beta0) based on the [`2.2.0` branch](https://github.com/jglev/obsidian-paste-to-current-indentation/compare/2.2.0#diff-564c860307257bfac9176f554a35aa858c350a59d75c0dcd0a5d63a480805bd0) just now. Would you be willing to try this out, and see whether it addresses the use-case you had in mind?

**Comment by anselmwang** [comment]
All the `paste-as` commands works perfectly!

I find two issues:
- When I set Paste Mode to be "Passthrough" and paste an image, nothing got pasted. In the past, a embed link like `![[attachments/Pasted image 20220407143235.png]]` appears.
- Now the plugin failed to load on Android.

**Comment by jglev** [comment]
Thank you for writing those up!

I _think_ that I've addressed both just now, in [release `2.2.0-beta1`](https://github.com/jglev/obsidian-paste-to-current-indentation/releases/tag/2.2.0-beta1) -- would you be willing to try that, as well? Thank you!

**Comment by jglev** [comment]
Sorry, I just realized that something seems to have gone wrong with a dependency in that last beta release; I'll re-post an updated version to try soon!

**Comment by jglev** [comment]
I've posted [`2.2.0-beta2`](https://github.com/jglev/obsidian-paste-to-current-indentation/releases/tag/2.2.0-beta2) just now to try out, if you're willing and available!

**Comment by anselmwang** [comment]
@jglev , now pasting image under "Passthrough" mode works. 

On the Android, all the "paste in xx mode" command doesn't work.

Previously in my implementation, "paste in text" and "paste in code" works and "paste in markdown" fails.

I don't have a conclusion but have a guess. On Android, when I run "clipboard.read()" for non-text format (i.e. html format), I will meet a "read permission denied" error. My "paste in text" and "paste in code" implementation only relies on "clipboard.readText()", so they works.  But current implementation loops through every format and might fail.

If that is the case, may be we need to try - catch - continue logic here.

**Comment by jglev** [comment]
@anselmwang, ah, sorry about that. I found the bug in the latest beta release (there was a promise-related race condition), and think that I've addressed it [in a new beta release](https://github.com/jglev/obsidian-paste-to-current-indentation/releases/tag/2.2.0-beta3).

**Comment by anselmwang** [comment]
@jglev , the plugin still doesn't work on mobile. I have debugged the code and it failed at "const clipboardData = await navigator.clipboard.read();", the error message is "NotAllowedError: Read Permission Denied".

BTW, in package.json, I need to add following lines to devDependencies to build the project
```
    "builtin-modules": "^3.2.0",
    "yarn": "^1.22.18"
```

**Comment by jglev** [comment]
@anselmwang, thank you for all that. I will work on debugging this further -- To confirm, I am not seeing that permissions error on Obsidian Mobile in Android on my device, so it isn't clear to me that `navigator.clipboard.read()` is necessarily the cause of the behavior you're seeing (vs. permissions changes or something similar in a specific version of Android or Obsidian Mobile).

Would you be willing to post the version of Obsidian Mobile and of Android that you're seeing this behavior with?

**Comment by jglev** [comment]
@anselmwang, I realize now that I am seeing the same permissions error -- I hadn't noticed it before when testing, and had thought that all commands were working on my device. I'm sorry for my confusion. I'm looking at #14; thank you for the suggestion.

**Comment by jglev** [comment]
@anselmwang, thank you again for your comments and your draft in #14. I've released a new beta, [`2.2.0-beta4`](https://github.com/jglev/obsidian-paste-to-current-indentation/releases/tag/2.2.0-beta4), just now. I took a general approach you had suggested (testing clipboard functionality using a `try/catch` block), but have gone the route of disabling the new "Paste in Markdown Mode" and "Paste in Markdown (Blockquote) Mode" commands when clipboard access doesn't support them (as on mobile). Pasting in those modes outside of the "Paste in... Mode" commands is still supported on mobile.

If you find any additional undesirable behavior from this beta, I'd welcome knowing about it, of course! Otherwise, I'm planning to test this for a few days and then hopefully feel comfortable releasing it this week.
